### PR TITLE
Make useAuthenticatedEndpoint configurable via CasC

### PR DESCRIPTION
**Pull request by Smasherr**
To make `useAuthenticatedEndpoint` configurable via YAML and configuration-as-code-plugin it was needed to change the visibility of its setter in `GitLabConnectionConfig`. I also used the recommended technique for `GlobalConfiguration#configure` by calling `StaplerRequest#bindJSON` inside of it.

I added an integration test with CasC, which tests evaluating `useAuthenticatedEndpoint` in YAML. Adding this test required updating Jenkins dependency from 1.609.3 to 2.187 and a little refactoring due to changes in `jenkins.model`.

### CODE REVIEW

File: pom.xml
Comment (oleg-nenashev) [inline pom.xml]: Not going to work that way. It should support few recent LTS baselines at least. I am not sure what is @omehegan 's preference, but I kindly request 2.150.x at least

File: pom.xml
Comment (oleg-nenashev) [inline pom.xml:144]: Not needed for newer developer tools. You can bump plugin POm to the recent version and use the `hpi.compatibleSinceVersion` property

File: pom.xml
Comment (oleg-nenashev) [inline pom.xml]: I suggest using 1.27 at least so that security fixes are aplied

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java
Comment (oleg-nenashev) [inline src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java:519]: unrelated

File: src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnectionConfig.java
Comment (oleg-nenashev) [inline src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnectionConfig.java]: Unrelated?

File: src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnectionConfig.java
Comment (oleg-nenashev) [inline src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnectionConfig.java:69]: Usual JCasC compatibility practice is to use `@DataBoundSetter` and `@DataBoundConstructor` for global configurations

File: src/test/java/com/dabsquared/gitlabjenkins/casc/GitLabConnectionConfigTest.java
Comment (oleg-nenashev) [inline src/test/java/com/dabsquared/gitlabjenkins/casc/GitLabConnectionConfigTest.java]: Might be better to use the test rules from the JCasC plugin

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
From what I see it is a partial fix at best. It might be acceptable as is if the recent LTS versions are supported, but I am not sure. Up to @omehegan as a maintainer

**Comment by oleg-nenashev** [comment]
CC https://github.com/jenkinsci/configuration-as-code-plugin/issues/809

File: src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java
Comment (Smasherr) [reply src/main/java/com/dabsquared/gitlabjenkins/GitLabPushTrigger.java:519]: This is what I am talking about in my opening comment. The return type of `jenkins.model.ParameterizedJobMixIn.ParameterizedJob#getTriggers` has changed that's why I need to do this change if I update dependency to jenkins-core. Otherwise I get the following compile error:
```Error:(519, 52) java: incompatible types: java.lang.Object cannot be converted to hudson.triggers.Trigger```

File: src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnectionConfig.java
Comment (Smasherr) [reply src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnectionConfig.java]: Reverted

File: pom.xml
Comment (Smasherr) [reply pom.xml:144]: Does it look better now?

**Comment by oleg-nenashev** [comment]
Test fails due to the unsatisfied plugin dependency.
https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fgitlab-plugin/detail/PR-965/2/tests/

```
Caused by: java.io.IOException: Pipeline: Execution Support v1.15 failed to load.
- Script Security Plugin v1.13 is older than required. To fix, install v1.16 or later.
```

I'f guess you want to just update the dependency

**Comment by oleg-nenashev** [comment]
Plugin POM is quite obsolete in this plugin.
It needs to be updated to the recent versions so that upper bounds checks start working,
In fact, 2.29 looks to be used deliberately so the enforcer is not invoked during the build

**Comment by Smasherr** [comment]
> Test fails due to the unsatisfied plugin dependency.
> https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fgitlab-plugin/detail/PR-965/2/tests/
> 
> ```
> Caused by: java.io.IOException: Pipeline: Execution Support v1.15 failed to load.
> - Script Security Plugin v1.13 is older than required. To fix, install v1.16 or later.
> ```
> 
> I'f guess you want to just update the dependency

Well I already tried to do that and that resulted in further plugin dependency issues. I'll demonstrate it.

File: pom.xml
Comment (Smasherr) [reply pom.xml:144]: @oleg-nenashev ping

Comment (oleg-nenashev) [reply pom.xml:144]: No. You need a very recent Plugin POM. You still have a version which is few years old

**Comment by oleg-nenashev** [comment]
> Well I already tried to do that and that resulted in further plugin dependency issues. I'll demonstrate it.


This is a valid behavior. You can always add a `enforcer.skip` property to ignore the enforcer rules, but  it is not exactly the best practice. I am not sure it will be considered by the maintainers

**Comment by omehegan** [comment]
@Smasherr unfortunately I have not had any time to work on this plugin for the past few months, I am not even keeping up with GitHub updates for the time being. If @oleg-nenashev can advise on best practices, and you can implement them, I would be happy to merge.

**Comment by markjacksonfishing** [comment]
I will handle updating this @oleg-nenashev @Smasherr

**Comment by Smasherr** [comment]
> I will handle updating this @oleg-nenashev @Smasherr

Could you please reference my PR in yours so I get notified?

**Comment by markjacksonfishing** [comment]
@Smasherr Is there an associated Github issue with this PR? What about a Jenkins jira? If so, can you link them? If not, can you open them and note this PR?
Thank you for that.

**Comment by Smasherr** [comment]
> @Smasherr Is there an associated Github issue with this PR? What about a Jenkins jira? If so, can you link them? If not, can you open them and note this PR?
> Thank you for that.

Here you go:

https://github.com/jenkinsci/gitlab-plugin/issues/974
https://issues.jenkins-ci.org/browse/JENKINS-59295

**Comment by Smasherr** [comment]
@markyjackson-taulia How's the progress, can I somehow support you?

**Comment by Smasherr** [comment]
@markyjackson-taulia Hi, any news on this topic?

**Comment by timja** [comment]
Can this PR be closed? 
A smaller fix was merged in https://github.com/jenkinsci/gitlab-plugin/pull/866

**Comment by Smasherr** [comment]
> Can this PR be closed?
> A smaller fix was merged in #866

It seems so, yes.

**Pull request closed without merging**
